src/workerd/jsg/struct.h

```diff
@@ -12,8 +12,69 @@
 #include <workerd/jsg/value.h>
 #include <workerd/jsg/web-idl.h>
 
+#include <concepts>
+#include <type_traits>
+
 namespace workerd::jsg {
 
+template <typename T>
+constexpr bool isV8LocalOrData = isV8Local<T>() || std::is_base_of_v<v8::Data, T> || IsJsValue<T>;
+
+template <typename T>
+constexpr bool isV8LocalOrData<kj::Maybe<T>> = isV8LocalOrData<T>;
+
+template <typename T>
+constexpr bool isV8LocalOrData<Optional<T>> = isV8LocalOrData<T>;
+
+template <typename T>
+constexpr bool isV8LocalOrData<LenientOptional<T>> = isV8LocalOrData<T>;
+
+template <typename T>
+constexpr bool isV8LocalOrData<kj::Array<T>> = isV8LocalOrData<T>;
+
+template <typename T>
+constexpr bool isV8LocalOrData<kj::ArrayPtr<T>> = isV8LocalOrData<T>;
+
+template <typename T>
+constexpr bool isV8LocalOrData<Dict<T>> = isV8LocalOrData<T>;
+
+template <typename T, typename... Rest>
+constexpr bool isV8LocalOrData<std::tuple<T, Rest...>> =
+    isV8LocalOrData<T> || isV8LocalOrData<std::tuple<Rest...>>;
+
+template <typename T, typename... Rest>
+constexpr bool isV8LocalOrData<kj::OneOf<T, Rest...>> =
+    isV8LocalOrData<T> || isV8LocalOrData<std::tuple<Rest...>>;
```

Comment:
I think this would avoid the `std::tuple` specialization, which we otherwise don't need:
```suggestion
    isV8LocalOrData<T> || (isV8LocalOrData<Rest> || ...);
```

Reply:
nice